File: src/main/cpp/subsystems/rollers/RollerSystem.cpp
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

#include "frc2025/subsystems/rollers/RollerSystem.h"

#include <frc/Alert.h>
#include <frc2/command/Command.h>
#include <frc2/command/SubsystemBase.h>
#include <wpi/function.h>

#include "frc2025/Robot.h"
#include "frc2025/util/LoggedTracer.h"
#include "third_party/frc-wpi-util/wpi/StringRef.h"
#include "third_party/frc-wpi-util/wpi/function_ref.h"
#include "third_party/frc-wpi-util/wpi/span.h"
#include "third_party/frc-wpi-util/wpi/string.h"
#include "third_party/frc-wpi-util/wpi/units/voltage.h"

using namespace frc;
using namespace frc2;
using namespace wpi;

RollerSystem::RollerSystem(std::string name, RollerSystemIO *io)
    : SubsystemBase(), name_(name), io_(io),
      disconnected_(name + " motor disconnected!", Alert::AlertType::kWarning),
      tempFault_(name + " motor too hot! \xF0\x9F\x98\xB5",
                 Alert::AlertType::kWarning) {}

void RollerSystem::Periodic() {
  io_->UpdateInputs(&inputs_);
  Logger::ProcessInputs(name_, &inputs_);
  disconnected_.Set(!inputs_.data.connected && !Robot::IsJitting());
  tempFault_.Set(inputs_.data.tempFault);

  // Record cycle time
  LoggedTracer::Record(name_);
}

CommandPtr RollerSystem::RunRoller(std::function<double()> inputVolts) {
  return StartEnd([this, inputVolts] { io_->RunVolts(inputVolts()); },
                  [this] { io_->Stop(); });
}

File: src/main/java/org/littletonrobotics/frc2025/subsystems/rollers/RollerSystem.java
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

package org.littletonrobotics.frc2025.subsystems.rollers;

import edu.wpi.first.wpilibj.Alert;
import lombok.Setter;
import org.littletonrobotics.frc2025.Robot;
import org.littletonrobotics.frc2025.util.LoggedTracer;
import org.littletonrobotics.junction.Logger;

public class RollerSystem {
  private final String name;
  private final String inputsName;
  private final RollerSystemIO io;
  protected final RollerSystemIOInputsAutoLogged inputs = new RollerSystemIOInputsAutoLogged();
  private final Alert disconnected;
  private final Alert tempFault;

  @Setter private double volts = 0.0;
  private boolean brakeModeEnabled = true;

  public RollerSystem(String name, String inputsName, RollerSystemIO io) {
    this.name = name;
    this.inputsName = inputsName;
    this.io = io;

    disconnected = new Alert(name + " motor disconnected!", Alert.AlertType.kWarning);
    tempFault = new Alert(name + " motor too hot! 🥵", Alert.AlertType.kWarning);
  }

  public void periodic() {
    io.updateInputs(inputs);
    Logger.processInputs(inputsName, inputs);
    disconnected.set(!inputs.data.connected() && !Robot.isJITing());
    tempFault.set(inputs.data.tempFault());

    // Run roller
    io.runVolts(volts);

    // Record cycle time
    LoggedTracer.record(name);

    Logger.recordOutput(inputsName + "/BrakeModeEnabled", brakeModeEnabled);
  }

  public void setBrakeMode(boolean enabled) {
    if (brakeModeEnabled == enabled) return;
    brakeModeEnabled = enabled;
    io.setBrakeMode(enabled);
  }

  public double getTorqueCurrent() {
    return inputs.data.torqueCurrentAmps();
  }

  public double getVelocity() {
    return inputs.data.velocityRadsPerSec();
  }

  public void stop() {
    volts = 0.0;
  }
}
